### PR TITLE
8342303: Segfault in update_inherited_vtable: AppCDS, old bytecode, and redefineClasses

### PR DESCRIPTION
**Pull request by calvinccheung**
A simple fix to avoid referencing invalid vtable and itable entires in `InstanceKlass::restore_unshareable_info` if a class is not linked during CDS dump time.

Passed tiers 1 - 4 testing.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8342303](https://bugs.openjdk.org/browse/JDK-8342303): Segfault in update_inherited_vtable: AppCDS, old bytecode, and redefineClasses (**Bug** - P4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21667/head:pull/21667` \
`$ git checkout pull/21667`

Update a local copy of the PR: \
`$ git checkout pull/21667` \
`$ git pull https://git.openjdk.org/jdk.git pull/21667/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21667`

View PR using the GUI difftool: \
`$ git pr show -t 21667`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21667.diff">https://git.openjdk.org/jdk/pull/21667.diff</a>

</details>

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8342303](https://bugs.openjdk.org/browse/JDK-8342303): Segfault in update_inherited_vtable: AppCDS, old bytecode, and redefineClasses (**Bug** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**) Review applies to [d1801774](https://git.openjdk.org/jdk/pull/21667/files/d180177489299d45b7882b0ab267d9e9ffeb7a4a)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**) Review applies to [d1801774](https://git.openjdk.org/jdk/pull/21667/files/d180177489299d45b7882b0ab267d9e9ffeb7a4a)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21667/head:pull/21667` \
`$ git checkout pull/21667`

Update a local copy of the PR: \
`$ git checkout pull/21667` \
`$ git pull https://git.openjdk.org/jdk.git pull/21667/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21667`

View PR using the GUI difftool: \
`$ git pr show -t 21667`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21667.diff">https://git.openjdk.org/jdk/pull/21667.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21667#issuecomment-2442699974)
</details>

### CODE REVIEW

File: test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/OldClassAndRedefineClassApp.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import java.lang.instrument.ClassDefinition;
+import java.lang.instrument.Instrumentation;
+
+public class OldClassAndRedefineClassApp {
+
+    public static void main(String args[]) throws Throwable {
+
+        System.out.println("Main: loading OldSuper");
+        // Load an old class (version 49), but not linking it.
+        Thread.currentThread().getContextClassLoader().loadClass("OldSuper");
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/OldClassAndRedefineClassApp.java:34]: Why the context classloader? Has it been set in this test?

**Review by dholmes-ora — COMMENTED** [review]
Functional fix looks good. Unclear about the the test details ...

Thanks

**Comment by calvinccheung** [comment]
I've added two tests:

- redefine a super class loaded by the boot class loader;
- redefine an old super class loaded by the AppClassLoader.

Also simplified the OldClassAndRedefineClass.java test.

File: test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/OldClassAndRedefineClassApp.java (hunk 1 above)
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/OldClassAndRedefineClassApp.java:34]: You kind of avoided the question though - why are you using the context classloader instead of e.g the classloader of the current class? Ther CCL is defined as follows:
```
Returns the context ClassLoader for this thread. The context ClassLoader may be set by the creator of the thread for use by code running in this thread when loading classes and resources. If not [set](https://docs.oracle.com/en/java/javase/22/docs/api/java.base/java/lang/Thread.html#setContextClassLoader(java.lang.ClassLoader)), the default is to inherit the context class loader from the parent thread.

The context ClassLoader of the primordial thread is typically set to the class loader used to load the application.
```
So if nothing sets it then you are just using the App/System loader, so why not just use that loader directly?

Comment (calvinccheung) [reply test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/OldClassAndRedefineClassApp.java:34]: I've updated the tests to avoid using `getContextClassLoader()loadClass()`.
For the App/System loader case:
    `ClassLoader.getSystemClassLoader().loadClass()`
For the boot loader case:
    `Class.forName(<class name>, false, null)`

File: test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/OldClassAndRedefineClassApp.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    public static void main(String args[]) throws Throwable {
+
+        System.out.println("Main: loading OldSuper");
+        // Load an old class (version 49), but not linking it.
+        ClassLoader.getSystemClassLoader().loadClass("OldSuper");
+
+        // Redefine a class unrelated to the above old class.
+        Instrumentation instrumentation = InstrumentationRegisterClassFileTransformer.getInstrumentation();
+        System.out.println("INFO: instrumentation = " + instrumentation);
+        Class c = ClassLoader.getSystemClassLoader().loadClass("Hello");
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/OldClassAndRedefineClassApp.java:39]:
```suggestion
        Class<?> c = ClassLoader.getSystemClassLoader().loadClass("Hello");
```

File: test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/OldClassAndRedefineClassApp.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import java.lang.instrument.ClassDefinition;
+import java.lang.instrument.Instrumentation;
+
+public class OldClassAndRedefineClassApp {
+
+    public static void main(String args[]) throws Throwable {
+
+        System.out.println("Main: loading OldSuper");
+        // Load an old class (version 49), but not linking it.
+        ClassLoader.getSystemClassLoader().loadClass("OldSuper");
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/OldClassAndRedefineClassApp.java:34]: Why not `Class.forName` in all cases ??

File: test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/RedefineBootClassApp.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+import java.io.File;
+import java.lang.instrument.ClassDefinition;
+import java.lang.instrument.Instrumentation;
+
+public class RedefineBootClassApp {
+    public static void main(String args[]) throws Throwable {
+        File bootJar = new File(args[0]);
+
+        Class superCls = Class.forName("BootSuper", false, null);
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/RedefineBootClassApp.java:33]:
```suggestion
        Class<?> superCls = Class.forName("BootSuper", false, null);
```

File: test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/RedefineBootClassApp.java [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        {
+            BootSuper obj = (BootSuper)superCls.newInstance();
+            String s = obj.doit();
+            System.out.println("(after transform) BootSuper>> doit() = " + s);
+            if (!s.equals("HELLO")) {
+                throw new RuntimeException("BootSuper doit() should be HELLO but got " + s);
+            }
+        }
+
+        Class childCls = Class.forName("BootChild", false, null);
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/RedefineBootClassApp.java:56]:
```suggestion
        Class<?> childCls = Class.forName("BootChild", false, null);
```

File: test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/RedefineOldSuperApp.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+import java.io.File;
+import java.lang.instrument.ClassDefinition;
+import java.lang.instrument.Instrumentation;
+
+public class RedefineOldSuperApp {
+    public static void main(String args[]) throws Throwable {
+        File bootJar = new File(args[0]);
+
+        Class superCls = ClassLoader.getSystemClassLoader().loadClass("OldSuper");
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/RedefineOldSuperApp.java:33]:
```suggestion
        Class<?> superCls = ClassLoader.getSystemClassLoader().loadClass("OldSuper");
```

File: test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/RedefineOldSuperApp.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        {
+            OldSuper obj = (OldSuper)superCls.newInstance();
+            String s = obj.doit();
+            System.out.println("(after transform) OldSuper>> doit() = " + s);
+            if (!s.equals("HELLO")) {
+                throw new RuntimeException("OldSuper doit() should be HELLO but got " + s);
+            }
+        }
+
+        Class childCls = ClassLoader.getSystemClassLoader().loadClass("NewChild");
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/RedefineOldSuperApp.java:56]:
```suggestion
        Class<?> childCls = ClassLoader.getSystemClassLoader().loadClass("NewChild");
```

**Review by dholmes-ora — CHANGES_REQUESTED** [review]

File: test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/OldClassAndRedefineClassApp.java (hunk 2 above)
Comment (calvinccheung) [reply test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/OldClassAndRedefineClassApp.java:39]: Fixed.

File: test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/OldClassAndRedefineClassApp.java (hunk 3 above)
Comment (calvinccheung) [reply test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/OldClassAndRedefineClassApp.java:34]: I've changed it to use Class.forName().

File: test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/RedefineBootClassApp.java (hunk 4 above)
Comment (calvinccheung) [reply test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/RedefineBootClassApp.java:33]: Fixed.

File: test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/RedefineBootClassApp.java (hunk 5 above)
Comment (calvinccheung) [reply test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/RedefineBootClassApp.java:56]: Fixed.

File: test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/RedefineOldSuperApp.java (hunk 6 above)
Comment (calvinccheung) [reply test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/RedefineOldSuperApp.java:33]: Fixed.

File: test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/RedefineOldSuperApp.java (hunk 7 above)
Comment (calvinccheung) [reply test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/RedefineOldSuperApp.java:56]: Fixed.
Also, changed `ClassLoader.getSystemClassLoader().loadClass()` to `Class.forName()`.

File: test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/OldClassAndRedefineClassApp.java [hunk 8]
```diff
@@ -31,12 +31,12 @@ public static void main(String args[]) throws Throwable {
...
         System.out.println("Main: loading OldSuper");
         // Load an old class (version 49), but not linking it.
-        ClassLoader.getSystemClassLoader().loadClass("OldSuper");
+        Class.forName("OldSuper", false, ClassLoader.getSystemClassLoader());
 
         // Redefine a class unrelated to the above old class.
         Instrumentation instrumentation = InstrumentationRegisterClassFileTransformer.getInstrumentation();
         System.out.println("INFO: instrumentation = " + instrumentation);
-        Class c = ClassLoader.getSystemClassLoader().loadClass("Hello");
+        Class<?> c = ClassLoader.getSystemClassLoader().loadClass("Hello");
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/OldClassAndRedefineClassApp.java:39]: Can you use `Class.forName` here too.

Comment (calvinccheung) [reply test/hotspot/jtreg/runtime/cds/appcds/jvmti/redefineClasses/OldClassAndRedefineClassApp.java]: Missed that one. Fixed now.

**Review by coleenp — COMMENTED** [review]
There is a RedefineHelper class that hides all this instrumentation and agent code, and makes redefinition tests really simple.  I can't tell if you can use that from all the code in these tests. Examples are in the serviceability/jvmti/RedefineClasses tests.

**Comment by calvinccheung** [comment]
> There is a RedefineHelper class that hides all this instrumentation and agent code, and makes redefinition tests really simple. I can't tell if you can use that from all the code in these tests. Examples are in the serviceability/jvmti/RedefineClasses tests.

In fact, some CDS tests are also using RedefineClassHelper. It did simply the test code. I've updated the tests.

**Review by dholmes-ora — APPROVED** [review]
Nothing further from me.

Thanks

**Review by iklam — APPROVED** [review]
LGTM

**Review by coleenp — APPROVED** [review]
Looks good.  A bit hard to follow since it uses classes already existing in the test directory, but that's good.

**Comment by calvinccheung** [comment]
@iklam, @coleenp, @dholmes-ora,
I've noticed some failures from tier3 testing due to the jdk/test/lib/compiler/InMemoryJavaCompiler class not found. So I added the class to the "@build" command of the affected tests.
Please re-review. Probably one re-review is enough.
Thanks!

**Review by dholmes-ora — APPROVED** [review]
Okay on the updates

Thanks

**Comment by calvinccheung** [comment]
Thanks for the reviews!

/integrate
